File: ncod/master/api/device.py
"""设备API"""

from typing import Optional, List
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from ncod.master.services.device import device_service
from ncod.master.middleware.auth import get_current_user, require_permissions
from ncod.core.logger import setup_logger

logger = setup_logger("device_api")
router = APIRouter(prefix="/api/v1/devices")


class DeviceCreate(BaseModel):
    """创建设备请求"""

    name: str
    mac_address: str
    ip_address: Optional[str] = None
    description: Optional[str] = None
    organization_id: Optional[str] = None
    slave_id: Optional[str] = None


class DeviceUpdate(BaseModel):
    """更新设备请求"""

    name: Optional[str] = None
    mac_address: Optional[str] = None
    ip_address: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    organization_id: Optional[str] = None
    slave_id: Optional[str] = None


@router.post("")
@require_permissions(["device:create"])
async def create_device(request: DeviceCreate, user: dict = Depends(get_current_user)):
    """创建设备"""
    try:
        success, message, device = await device_service.create_device(request.dict())
        if not success:
            raise HTTPException(status_code=400, detail=message)
        return device
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating device: {e}")
        raise HTTPException(status_code=500, detail="Failed to create device")


@router.put("/{device_id}")
@require_permissions(["device:update"])
async def update_device(
    device_id: str, request: DeviceUpdate, user: dict = Depends(get_current_user)
):
    """更新设备"""
    try:
        success, message, device = await device_service.update_device(
            device_id, request.dict(exclude_unset=True)
        )
        if not success:
            raise HTTPException(status_code=400, detail=message)
        return device
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating device: {e}")
        raise HTTPException(status_code=500, detail="Failed to update device")


@router.delete("/{device_id}")
@require_permissions(["device:delete"])
async def delete_device(device_id: str, user: dict = Depends(get_current_user)):
    """删除设备"""
    try:
        success, message = await device_service.delete_device(device_id)
        if not success:
            raise HTTPException(status_code=400, detail=message)
        return {"message": "Device deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting device: {e}")
        raise HTTPException(status_code=500, detail="Failed to delete device")


@router.get("/{device_id}")
@require_permissions(["device:read"])
async def get_device(device_id: str, user: dict = Depends(get_current_user)):
    """获取设备"""
    try:
        device = await device_service.get_device(device_id)
        if not device:
            raise HTTPException(status_code=404, detail="Device not found")
        return device
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting device: {e}")
        raise HTTPException(status_code=500, detail="Failed to get device")


@router.get("")
@require_permissions(["device:read"])
async def list_devices(
    organization_id: Optional[str] = None,
    slave_id: Optional[str] = None,
    user: dict = Depends(get_current_user),
):
    """获取设备列表"""
    try:
        return await device_service.list_devices(organization_id, slave_id)
    except Exception as e:
        logger.error(f"Error listing devices: {e}")
        raise HTTPException(status_code=500, detail="Failed to list devices")

File: ncod/frontend/src/api/device.ts
import { http } from '@/utils/http';

export interface DeviceCreate {
  name: string;
  mac_address: string;
  ip_address?: string;
  description?: string;
  organization_id?: string;
  slave_id?: string;
}

export interface DeviceUpdate {
  name?: string;
  mac_address?: string;
  ip_address?: string;
  description?: string;
  is_active?: boolean;
  organization_id?: string;
  slave_id?: string;
}

export const deviceApi = {
  // 创建设备
  createDevice(data: DeviceCreate) {
    return http.post('/api/v1/devices', data);
  },

  // 更新设备
  updateDevice(deviceId: string, data: DeviceUpdate) {
    return http.put(`/api/v1/devices/${deviceId}`, data);
  },

  // 删除设备
  deleteDevice(deviceId: string) {
    return http.delete(`/api/v1/devices/${deviceId}`);
  },

  // 获取设备
  getDevice(deviceId: string) {
    return http.get(`/api/v1/devices/${deviceId}`);
  },

  // 获取设备列表
  listDevices(params?: {
    organization_id?: string;
    slave_id?: string;
  }) {
    return http.get('/api/v1/devices', { params });
  }
}; 